File: app/src/main/java/com/example/recipeclipper/data/local/dao/PantryDao.kt
package com.example.recipeclipper.data.local.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import com.example.recipeclipper.data.local.entity.PantryItemEntity
import kotlinx.coroutines.flow.Flow

/** The pantry (#51). Sorting and searching are done in memory ([com.example.recipeclipper.data.model.PantryList]). */
@Dao
abstract class PantryDao {

    @Query("SELECT * FROM pantry_items ORDER BY id ASC")
    abstract fun observeItems(): Flow<List<PantryItemEntity>>

    @Query("SELECT * FROM pantry_items ORDER BY id ASC")
    abstract suspend fun items(): List<PantryItemEntity>

    @Query("SELECT * FROM pantry_items WHERE id = :id")
    abstract suspend fun item(id: Long): PantryItemEntity?

    @Insert
    abstract suspend fun insert(item: PantryItemEntity): Long

    @Query("UPDATE pantry_items SET inStock = :inStock, updatedAt = :now WHERE id IN (:ids)")
    abstract suspend fun setInStock(ids: List<Long>, inStock: Boolean, now: Long)

    /** Back in stock, bought on [day]. */
    @Query("UPDATE pantry_items SET inStock = 1, purchasedDay = :day, updatedAt = :now WHERE id IN (:ids)")
    abstract suspend fun restock(ids: List<Long>, day: Long, now: Long)

    @Query(
        "UPDATE pantry_items SET name = :name, quantity = :quantity, alwaysHave = :alwaysHave, " +
            "expiresDay = :expiresDay, updatedAt = :now WHERE id = :id"
    )
    abstract suspend fun edit(id: Long, name: String, quantity: String?, alwaysHave: Boolean, expiresDay: Long?, now: Long)

    @Query("DELETE FROM pantry_items WHERE id = :id")
    abstract suspend fun delete(id: Long)

    /** Undoes a delete, or a restock, or a stock change: the row exactly as it was. */
    @Insert(onConflict = androidx.room.OnConflictStrategy.REPLACE)
    abstract suspend fun put(items: List<PantryItemEntity>)
}

File: ios/RecipeClipper/Data/Local/PantryDao.swift
import Foundation

/// The pantry's SQL (#51), ported from Android's PantryDao. Synchronous; run inside
/// `AppDatabase.read` / `write` (a write is one transaction).
struct PantryDao {
    let db: SQLiteConnection

    func items() throws -> [PantryItemRecord] {
        try db.query("SELECT \(PantryItemRecord.columns) FROM pantry_items ORDER BY id ASC", map: PantryItemRecord.init(row:))
    }

    func item(_ id: Int64) throws -> PantryItemRecord? {
        try db.queryOne("SELECT \(PantryItemRecord.columns) FROM pantry_items WHERE id = ?", id, map: PantryItemRecord.init(row:))
    }

    /// An `id` of 0 lets SQLite assign one; a non-zero id is honoured (what undo relies on).
    @discardableResult
    func insert(_ e: PantryItemRecord) throws -> Int64 {
        try db.run(
            "INSERT INTO pantry_items (\(PantryItemRecord.columns)) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            e.id == 0 ? nil : e.id, e.name, e.quantity, e.language, e.aisle, e.inStock, e.alwaysHave,
            e.purchasedDay, e.expiresDay, e.updatedAt, e.uid
        )
        return db.lastInsertRowId
    }

    func setInStock(_ ids: [Int64], inStock: Bool, now: Int64) throws {
        for id in ids {
            try db.run("UPDATE pantry_items SET inStock = ?, updatedAt = ? WHERE id = ?", inStock, now, id)
        }
    }

    /// Back in stock, bought on `day`.
    func restock(_ ids: [Int64], day: Int64, now: Int64) throws {
        for id in ids {
            try db.run("UPDATE pantry_items SET inStock = 1, purchasedDay = ?, updatedAt = ? WHERE id = ?", day, now, id)
        }
    }

    func edit(_ id: Int64, name: String, quantity: String?, alwaysHave: Bool, expiresDay: Int64?, now: Int64) throws {
        try db.run(
            "UPDATE pantry_items SET name = ?, quantity = ?, alwaysHave = ?, expiresDay = ?, updatedAt = ? WHERE id = ?",
            name, quantity, alwaysHave, expiresDay, now, id
        )
    }

    func delete(_ id: Int64) throws {
        try db.run("DELETE FROM pantry_items WHERE id = ?", id)
    }

    /// Undoes a delete, a restock or a stock change: the rows exactly as they were.
    func put(_ items: [PantryItemRecord]) throws {
        for item in items {
            try db.run("DELETE FROM pantry_items WHERE id = ?", item.id)
            try insert(item)
        }
    }
}
